File: akka-sample-persistence-dc-scala/src/main/scala/sample/persistence/res/counter/ThumbsUpCounter.scala
package sample.persistence.res.counter

import akka.actor.typed.scaladsl.Behaviors
import akka.actor.typed.{ActorRef, Behavior}
import akka.cluster.sharding.typed.ReplicatedEntityProvider
import akka.persistence.cassandra.query.scaladsl.CassandraReadJournal
import akka.persistence.typed.scaladsl.{Effect, EventSourcedBehavior, ReplicatedEventSourcing}
import akka.persistence.typed.{ReplicaId, ReplicationId}
import sample.persistence.res.{CborSerializable, MainApp}

object ThumbsUpCounter {

  // sent over sharding
  sealed trait Command extends CborSerializable {
    def resourceId: String
  }

  final case class GiveThumbsUp(resourceId: String, userId: String, replyTo: ActorRef[Long]) extends Command
  final case class GetCount(resourceId: String, replyTo: ActorRef[Long]) extends Command
  final case class GetUsers(resourceId: String, replyTo: ActorRef[State]) extends Command

  // saved to DB
  sealed trait Event extends CborSerializable
  final case class GaveThumbsUp(userId: String) extends Event

  // saved to DB
  final case class State(users: Set[String]) extends CborSerializable {
    def add(userId: String): State = copy(users + userId)
  }

  val Provider: ReplicatedEntityProvider[Command] = ReplicatedEntityProvider.perDataCenter("counter", MainApp.AllReplicas) { replicationId => ThumbsUpCounter(replicationId) }

  // we use a shared journal as cassandra typically spans DCs rather than a DB per replica
  def apply(replicationId: ReplicationId): Behavior[Command] =
    Behaviors.setup { ctx =>
      ReplicatedEventSourcing.commonJournalConfig(replicationId, Replicas, CassandraReadJournal.Identifier) { replicationContext =>
        EventSourcedBehavior[Command, Event, State](
          persistenceId = replicationId.persistenceId,
          emptyState = State(Set.empty),
          commandHandler = (state, cmd) => cmd match {
            case GiveThumbsUp(_, userId, replyTo) =>
              Effect.persist(GaveThumbsUp(userId)).thenRun { state2 =>
                ctx.log.info("Thumbs-up by {}, total count {}", userId, state2.users.size)
                replyTo ! state2.users.size
              }
            case GetCount(_, replyTo) =>
              replyTo ! state.users.size
              Effect.none
            case GetUsers(_, replyTo) =>
              replyTo ! state
              Effect.none
          },
          eventHandler = (state, event) => event match {
            case GaveThumbsUp(userId) =>
              state.add(userId)
          }
        )
      }
    }
}

File: akka-sample-persistence-dc-java/src/main/java/sample/persistence/res/counter/ThumbsUpCounter.java
package sample.persistence.res.counter;

import akka.actor.typed.ActorRef;
import akka.actor.typed.Behavior;
import akka.actor.typed.javadsl.ActorContext;
import akka.actor.typed.javadsl.Behaviors;
import akka.cluster.sharding.typed.ReplicatedEntityProvider;
import akka.persistence.cassandra.query.javadsl.CassandraReadJournal;
import akka.persistence.typed.ReplicationId;
import akka.persistence.typed.javadsl.*;
import com.fasterxml.jackson.annotation.JsonCreator;
import sample.persistence.res.CborSerializable;

import java.util.Collections;
import java.util.HashSet;
import java.util.Set;

import static sample.persistence.res.MainApp.ALL_REPLICAS;

public final class ThumbsUpCounter extends ReplicatedEventSourcedBehavior<ThumbsUpCounter.Command, ThumbsUpCounter.Event, ThumbsUpCounter.State> {
   
    private final ActorContext<Command> ctx;

    private static Behavior<Command> create(ReplicationId id) {
        return Behaviors.setup(ctx -> ReplicatedEventSourcing.commonJournalConfig(id, ALL_REPLICAS, CassandraReadJournal.Identifier(), replicationContext -> new ThumbsUpCounter(replicationContext, ctx)));
    }

    public static ReplicatedEntityProvider<Command> provider() {
        return ReplicatedEntityProvider.createPerDataCenter(Command.class, "counter", ALL_REPLICAS, ThumbsUpCounter::create);
    }

    private ThumbsUpCounter(ReplicationContext replicationContext, ActorContext<Command> ctx) {
        super(replicationContext);
        this.ctx = ctx;
    }

    @Override
    public State emptyState() {
        return new State();
    }

    @Override
    public CommandHandler<Command, Event, State> commandHandler() {
        return newCommandHandlerBuilder().forAnyState()
                .onCommand(GiveThumbsUp.class, (state, cmd) -> Effect().persist(new GaveThumbsUp(cmd.userId)).thenRun(state2 -> {
                    ctx.getLog().info("Thumbs-up by {}, total count {}", cmd.userId, state2.users.size());
                    cmd.replyTo.tell((long) state2.users.size());
                })).onCommand(GetCount.class, (state, cmd) -> {
                    cmd.replyTo.tell((long) state.users.size());
                    return Effect().none();
                }).onCommand(GetUsers.class, (state, cmd) -> {
                    cmd.replyTo.tell(state);
                    return Effect().none();
                }).build();
    }

    @Override
    public EventHandler<State, Event> eventHandler() {
        return newEventHandlerBuilder().forAnyState()
                .onEvent(GaveThumbsUp.class, (state, event) ->
                        state.add(event.userId)
                ).build();

    }

    // Classes for commands, events, and state...

    public interface Command extends CborSerializable {
    }

    public static class GiveThumbsUp implements Command {
        public final String resourceId;
        public final String userId;
        public final ActorRef<Long> replyTo;

        public GiveThumbsUp(String resourceId, String userId, ActorRef<Long> replyTo) {
            this.resourceId = resourceId;
            this.userId = userId;
            this.replyTo = replyTo;
        }
    }

    public static class GetCount implements Command {
        public final String resourceId;
        public final ActorRef<Long> replyTo;

        public GetCount(String resourceId, ActorRef<Long> replyTo) {
            this.resourceId = resourceId;
            this.replyTo = replyTo;
        }
    }

    public static class GetUsers implements Command {
        public final String resourceId;
        public final ActorRef<State> replyTo;

        public GetUsers(String resourceId, ActorRef<State> replyTo) {
            this.resourceId = resourceId;
            this.replyTo = replyTo;
        }
    }


    interface Event extends CborSerializable {
    }

    public static class GaveThumbsUp implements Event {
        public final String userId;

        @JsonCreator
        public GaveThumbsUp(String userId) {
            this.userId = userId;
        }
    }

    public static class State implements CborSerializable {
        public final Set<String> users;

        public State() {
            this.users = Collections.emptySet();
        }

        State(Set<String> users) {
            this.users = Collections.unmodifiableSet(users);
        }

        public State add(String userId) {
            // defensive copy for thread safety
            Set<String> newUsers = new HashSet<>(users);
            newUsers.add(userId);
            return new State(newUsers);
        }
    }

}

